### DISCUSSION TITLE
Multithread Support with Emscripten

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#110] [Q&A]
I saw the other discussion [HERE ](https://github.com/jrouwe/JoltPhysics.js/discussions/41)but that person is trying to implement their own pooling/situation without the use of Pthreads.

As [Emscripten supports pThreads](https://emscripten.org/docs/porting/pthreads.html) and [C++ Threads are fairly common in WebAssebly](https://web.dev/articles/webassembly-threads#c_2) I'm wondering what it would take to get this functionality in Jolt. 

From my limited understanding, `std::threads` is built on-top of `pThreads` so from other examples I've seen should compile with Emscripten without too much hassle. There are limitations and things that might have to be done but I don't think it as big of a task as that other user. 

From the [feature table](https://webassembly.org/features/) Threads are well supported in all major browsers, so this looks like a doable thing.

One thing to note, is regardless of thread support with Jolt, Web users would likely still be required to use workers to access things as the point with Jolt is we can access them INSIDE other threads. But this way at least, we would push a lot of the body/object tracking back into jolt vs doing it all in duplicate. 

I haven't seen anything in the jolt docs that would let you get like 4 threads going for collision detection vs 1, but that's probably due to my lack of C++ skills.

**Comment by PhoenixIllusion** [comment]
To enable a build with pThreads, currently you need to make sure that all 3x sub-projects are compiled with flags to enable shared memory and multi-threading.

The project builds a jolt-static library, so that needs it
```
Line 32: set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread -s SHARED_MEMORY")
# Include Jolt
``` 
The project also builds a glue.o, and then finally the wasm module again
Set it again on the EMCC_ARGS and EMCC_GLUE_ARGS
```
Line 62:   --closure=1 -pthread -s SHARED_MEMORY
Line 105: -std=c++17 -pthread -s SHARED_MEMORY
```

And set the environment to run in a worker:
```
Line 65:  -s ENVIRONMENT='web,node,worker'
```
You will need to copy the /dist/*.worker.js file to your normal JS folder so that the main thread can spawn the child-threads
By default, Jolt will startup and create `(int)thread::hardware_concurrency() - 1 ` workers

Note: Chrome is fine with loading this on localhost as long as you have COOP/COEP enabled
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer

The biggest blocker on enabling threads so far is Safari has problems with the SharedMemory access threads. This causes problems where suddenly collision stops working all together, so usually after running with pThread enabled, you will just have bodies stop responding to collision detection and fall through each other. I run the test every so often, and have been looking at new Safari Preview release notes
![2024-02-21-multi-thread-falling-demo](https://github.com/jrouwe/JoltPhysics.js/assets/756488/dc702483-0bd8-4088-8725-08fe9583e283)

Non-Preview (existing Safari, iPhone) - Safari 17.2.1, iPhone iOS 17.3.1
* Bad News: Still has objects fall through the floor

Safari Preview - Release 188 (Safari 17.4) - requires HTTPS even on localhost + COEP/COOP:
* Good News: Safari Preview doesn't fail horribly and have objects fall through the floor
* Bad News: Safari Preview sometimes becomes unresponsive after a couple seconds? May not be an issue once in Stable

![2024-02-21-multi-thread-falling-safari-preview](https://github.com/jrouwe/JoltPhysics.js/assets/756488/fbbe3bbe-c24f-4b34-bfb8-f0d7df0c9d1c)

The usage of custom threads and pools is because of the way that Javascript interfaces with the Jolts internal event system. To some degree it was also me seeing if I could hack my way around the Safari bug, but also was to give myself extra hooks into the Thread Spawn logic.
If a Javascript callback gets fired in one thread, we want to make sure that it jumps to the JS world and gets directed back to whatever Javascript thread registered it. This requires some level of coordination between the WASM threading and the Javascript side (which is a bit challenging with emscripten given it doesn't really tell you when it spawns or even how to access those workers outside its internals)  so that we can at least catch that threads are spun up and pass along our custom logic to those new JSWorker threads so that they can properly respond to Collisions/Validations/Etc


Even if it does become stable, you'd want to ensure that you actually get a benefit from running the physics in more than 1x thread. Even with single-threading the physics, you can still setup to use an external SharedMem blob to syncs Position/Rotations and run [UI Main Thread, 3D Render Thread, PhysicsThread] without multi-threading the Physics Library

**Reply by PhoenixIllusion** [reply]
To see some of the initial problems in the bottom issue described, run a multi-threaded build with the following added to the bottom of your examples (such as 2d_funnel)
```javascript
const contactListener = new Jolt.ContactListenerJS();
contactListener.OnContactValidate = (body1, body2, baseOffset, collideShapeResult) => {
	return Jolt.ValidateResult_AcceptAllContactsForThisBodyPair;
};
contactListener.OnContactAdded = (body1, body2, manifold, settings) => {};
contactListener.OnContactPersisted = (body1, body2, manifold, settings) => {};
contactListener.OnContactRemoved = (subShapePair) => {};
physicsSystem.SetContactListener(contactListener);
```

Upon running in debug, you will see errors thrown of `Module.getCache is not a function`. This is because by default pThread defined worker will start auto-running on-spawn (via run() -> PTHREAD_ENV -> startWorker method in `*.worker.js`), before they are able to declare the Jolt JS interface of the method [getCache] and skip almost all of the WebIDL binding. The non-pThread code has several extra steps defined that will allow runtime & WebIDL initializing and uses setTimeout calls to process preRun/doRun/postRun methods.

We can emulate that behavior by modifying the `jolt-physics.wasm-compat.worker.js` file `self.startWorker` method to trigger via a setTimeout

```javascript 
    self.startWorker = (instance) => {
      setTimeout(() => {
        Module = instance;
        // Notify the main thread that this thread has loaded.
        postMessage({ 'cmd': 'loaded' });
        // Process any messages that were queued before the thread was ready.
        for (let msg of messageQueue) {
          handleMessage(msg);
        }
        // Restore the real message handler.
        self.onmessage = handleMessage;
      })
    };
``` 

After this, the code works kind-of as expected, but all workers will throw the expected Errors due to `ContactListenerJS` being called in worker-threads, because the Javascript worker's version of the Module does not actually have the JS Objects defined and registered inside it needed to call [OnContactAdded, OnContactPersisted, OnContactRemoved, OnContactValidate]

```
Cannot read properties of undefined (reading 'hasOwnProperty') @ runEmAsmFunction
```

Currently this will fail almost all JS-callback dependent demos: PathConstraint, ConveyorBelt, Buoyancy, ContactListener
Vehicle ones seem fine for now, but that may just be because we only have 1x StepListener going.

**Reply by jrouwe** [reply]
I thought I'd make it a bit easier to try this out and add an ENABLE_MULTI_THREADING option to CMakeLists.txt following your instructions above (see [branch](https://github.com/jrouwe/JoltPhysics.js/tree/feature/multithreading)) but when I try to build it it fails in the very last step with:

```
[100%] Built target jolt-wasm
[BABEL] Note: The code generator has deoptimised the styling of /tmp/tmp76h2bxfv.cc.js as it exceeds the max of 500KB.
wasm2js: /b/s/w/ir/cache/builder/emscripten-releases/binaryen/src/passes/I64ToI32Lowering.cpp:382: void wasm::I64ToI32Lowering::visitLoad(Load *): Assertion `!curr->isAtomic && "64-bit atomic load not implemented"' failed.
emcc: error: '/home/jorrit/emsdk/upstream/bin/wasm2js --emscripten -O /mnt/c/Users/jrouw/Documents/Code/JoltPhysics.js/dist/jolt-physics.wasm --mvp-features --enable-threads --enable-bulk-memory --enable-mutable-globals --enable-sign-ext' failed (received SIGABRT (-6))
gmake[2]: *** [CMakeFiles/jolt-javascript.dir/build.make:75: ../../dist/jolt-physics.js] Error 1
```

I'm running:

```
./build.sh Distribution -DENABLE_MULTI_THREADING=YES
```

Any ideas?

**Reply by PhoenixIllusion** [reply]
@jrouwe 
Sorry, I'd forgotten I disable ASM.JS in my builds. The atomic operations needed for shared-mem multi-threading will only work for WASM and WASM-compat.

Technically it should be possible in regular JS, but the `wasm2js` doesn't appear to implement the 64-bit atomic operations as of now, and aborts those attempts. https://github.com/WebAssembly/binaryen/blob/main/src/passes/I64ToI32Lowering.cpp

**Comment by DennisSmolek** [comment]
Thank you for the super in-depth response! I appreciate your insights and all the work you've clearly done on this path. 

Regarding your main blocker, running in safari: 
> Non-Preview (existing Safari, iPhone) - Safari 17.2.1, iPhone iOS 17.3.1

Is this only an issue on Mobile safari or is the desktop version showing similar issues?

Have other platforms/orgs discussed similar issues with `SharedArrayBuffer`'s still remaining in Safari? The discussions I've read seem to refer to issues that cleared up a few years ago.

---

Obviously, I'd like to see everything working perfectly on every platform, but if Mobile Safari is the only platform with such issues then I think it not a bad idea to move forward with a working threaded version. We could disclaimer the limitation for Safari continue to test/work towards safari working properly. 

If we had a multi-thread version we could build examples for the different usecases/setups and deal with other issues like the ones you mentioned in your second reply. 

Once safari multi-threading became stable they could use the same examples/docs for their setup...

---

> Even if it does become stable, you'd want to ensure that you actually get a benefit from running the physics in more than 1x thread. Even with single-threading the physics, you can still setup to use an external SharedMem blob to syncs Position/Rotations and run [UI Main Thread, 3D Render Thread, PhysicsThread] without multi-threading the Physics Library

Totally agree. For me, it's adding/removing bodies to the system as well as syncing position data that would be most advantageous. Also having a single point to listen to contact/body events without having to pipe events through `postMessages` everywhere.

**Reply by PhoenixIllusion** [reply]
Safari on desktop also appears to have the same issues, though I need to OS update to get the very latest revision of Safari.

Those builds may want to be their own distro target, like wasm/wasm-compat, although I think they haven't been too unstable when launched with a max thread of 1x

Regarding the events, many collision listeners methods have a response, like contact-validate, or use ioParameters allowing modifications in the middle of a worker thread's Job, so you if those are used, you need to design in a way that the worker threads can correctly respond to those instead of just firing off an event. Likely you would not want to use postMessage and await a reply from a single logic processing thread or you've erased all the parallel benefits.

It is possible to never use any of those fields, in which case it's not an issue. Also, anything living on the C++ layer is automatically handed across threads. Ex: Collision Filter and Sensor flags can substitute for some (but not all) contact validation use cases. Compiling your own needed logic against the source is a viable solution in some special cases as well.

If custom JS classes are used, likely the ideal solution is to design such that you can replicate the exact custom collision logic to the workers, ex: serialize your constructors and pass them to the threads or use a shared buffer, so that all workers know how to correctly respond the exact same way given what limited logic they know.

**Reply by PhoenixIllusion** [reply]
I'd need to double check, but from my experience, without some custom Jolt C++ side job code, a pthread worker will never respond to a postMessage once it's started, since pThreads are a single routine that starts and never stops or pauses unless instructed from the C++ side.

This means for practical tasks you'd either need to build and pass everything over prior to initializing and running the first Step, or we'd need to design some custom Job task worker.

**Comment by DennisSmolek** [comment]
I didn’t mean to distract with my `postMessages` comment. I simply meant communicating with workers and the main thread requires a lot of handles and piping. By using Jolts internal usage of threads developers wouldn’t need such direct worker access..

**to keep this thread on track**, what is your test setup? I can run it against safari latest on desktop and mobile and see if I can replicate the results.

also I like the idea of multi-thread being a build target, what would the steps be to make that happen?

**Reply by PhoenixIllusion** [reply]
To build your own version, the reference branch from jrouwe [https://github.com/jrouwe/JoltPhysics.js/tree/feature/multithreading] works as-is, if you change line 26 to only build wasm-compat using the --target param to avoid the binareyn wasm2js atomics issue we mentioned:

```sh
cmake --build Build/$BUILD_TYPE -j`nproc` --target jolt-wasm-compat
```

Then run the command provided above.:
```sh
 sh build.sh Distribution -DENABLE_MULTI_THREADING=YES
```
You need to specify 'Distribution' as the arg2 and not just rely on it defaulting to Distribution, because it blindly takes arg2 as the script $BUILD_TYPE, and will build a CMake build literally named `Build/-DENABLE_MULTI_THREADIN=YES` , a strange looking build-folder name

You will need to copy the resulting `dist/*.worker.js` file into the Examples/js/ folder and you will need some way to self-host a server on your machine with the following COOP/COEP headers
```
 res.setHeader('Cross-Origin-Embedder-Policy', 'require-corp');
 res.setHeader('Cross-Origin-Opener-Policy', 'same-origin');
```

WIthout modifying any of the Examples, you can produce issues using the `conveyor_belt.html` example.

Regarding developers not needing access to workers, internally the WASM threads can access each others WASM/C++ memory, but nothing on each others JS layer, so Each thread will have it's own JS Listener callbacks trigger and would be expected by Jolt to independently handle and respond to them as required. Jolt expects from the C++ side that shared-mem means shared functions, so this logic is sound from a C++ viewpoint.

Regarding a kind of flexible solution to the worker-sync issue, I think we can subclass JobSystemThreadPool and create a `sync` method on it that uses a fixed `EM_JS` method,

```cpp
EM_JS( void, __synchronize_jolt_worker, (), { globalThis.synchronizeWorker && globalThis.synchronizeWorker() });
```
I'm not sure on the best logic or location of calling it, but possibly it can be called in the ThreadMain before the `Semaphore Acquire`. Each worker thread can then have user-specified methods tacked onto them or loaded during worker-init that uses this generic block to populate or update the JS layer as-needed prior to the jobs being run.

My PostMessage comment itself was just as a callout to anyone who uses Worker Threads in JS, that once the pThread starts and the thread-pool is running, the Thread is essentially locked in a while(true) loop and will never respond to you from the JS side, even with a postMessage, which JS developers may expect would work. Someone who uses VanillaJS may expect that as a trivial solve, they would just add a logic step every cycle to postMessage pass over any pointers from the main thread and ensure they sync everything to the workers before they run the next Physics System Step. But, if just using emscripten's default behavior, that strategy won't work.

**Reply by PhoenixIllusion** [reply]
I was able to update Safari and confirm the existing issue on Desktop - Desktop Safari 17.3.1, Released January 22, 2024

This is even with setting a 0-thread-count on it and spawning 0-workers, though Safari works fine if using the compiled single-threaded Jolt.

**Comment by jrouwe** [comment]
> Bad News: Safari Preview sometimes becomes unresponsive after a couple seconds? May not be an issue once in Stable

I think I have an idea why this happens, I had to increment the thread pool size and match it with the max threads that the job scheduler will take or else I got hangs in Chrome as well. Maybe you can try again with the new code and see if it is fixed now?

Anyway the change is now merged and when enabled it gives a very good speed boost (the stress test runs 15fps without it and 60fps with it on my machine).

B.t.w. I tried making a web server that enables COOP/COEP but this is googled together, let me know if I did it totally wrong (especially how you're supposed to work with packages is a mystery to me).

**Reply by PhoenixIllusion** [reply]
The new CMake/Build script appears to work end-to-end for me.

The only change I would recommend as a Node standard is to use either `start` or `serve` to launch it, and put that in the `package.json`. The description using `Serves`, so `serve` may make sense in that regard and it can be run via `npm run serve`. If you use `start` then it can be run via `npm start` or `npm run start` 

```json
  "description": "Serves the JoltPhysics.js examples",
  "main": "serveexamples.js",
  "scripts": {
    "serve": "node serveexamples.js"
  },
```

After that, this worked fine
```
sh build.sh Distribution -DENABLE_MULTI_THREADING=ON
cd serveexamples/
npm install
npm run serve
```

**Reply by PhoenixIllusion** [reply]
All the headers we need for COOP/COEP are showing for me as well:
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/84d5306a-d1d0-41ff-8b57-37cdcd65d40a)

The fact that CharacterVirtual is not multi-threaded itself since it runs outside the physics Step call at least gives some flexibility in character-driven games, and if you only ever care about *only* the character hitting things in callbacks, that route can work for an entire game, but the Vehicles dependency on JS physics callbacks fails any cars and motorcycles driving games. (without a fix on the workers)

**Reply by jrouwe** [reply]
Thanks for the tip!

Did the Safari Preview version work without hanging now as well?

**Reply by PhoenixIllusion** [reply]
I hadn't had too much of a problem with that prior to your modification on Safari Build 189, as long as it was a Distribution build with no bugs. Safari really doesn't do well with trying to load a debug instance or any other large JS files, even for regular front-end web development. The Safari Preview status seems to change every-other-week, given it is the rolling every-2-week webkit build (https://webkit.org/blog/). If there is ever a crash in a thread, likely you need to terminate the entire tab and restart another one from scratch because the entire process can lock up.

That also sometimes seems to be the case in single-threaded Chrome/Firefox sometimes as well, especially if a debug-build with how massive debug-binaries are, or if you frequently recompile a Distro build. I think as long as the wasm binary doesn't change, the browser will just use the cached bytecode compile in the process, but if you rebuilt the wasm module, the browser will recompile wasm-to-bytecode but not always free the old tab-process memory, until finally Jolt instantiates but is only give a 0-byte heap and mallocs will fail on soft reload of the page. Crashes in any WASM code seem to also leak memory in a tab-process over time, so frequent WASM crashes will also destabilize a browser-tab-process until it needs to be killed and restarted (duplicate-tab seems fine)

Parallelism:
In general, Shared-Mem+atomic threads are not mainstream yet, and all platforms are having some issues about the best way to maintain and clean them up without leaking in a process. In the wild, I've seen more heavily emphasis on just plain scatter/gather of longer running process with transferable buffers, as in break into chunks and run in parallel and recombine-on-main, and some emphasis on web-gpu, but the web-gpu is still mainly in trivial gpu-side simulation ports of non-web code or just pure graphics related ports. Safari is at least starting to re-enable WebGPU after they had to scrap all their prior web-gpu work, since they'd essentially done all the older work in Metal and only recently started convert over to the official language standard. Safari Preview just put the option to test WebGPU again 1 month ago with knowledge it will fail some percentage of tests.

**Comment by PhoenixIllusion** [comment]
A perk of the PTHREAD_POOL_SIZE logic is that it will pre-init all the threads, then pause until you actually create the physicsSystem. This gives both the ability to fully start up the thread and Jolt, and you almost have a solid barrier where you can configure the worker threads.

Taking that into account, when you do a multi-threaded JS run of Jolt, you can pass a custom script that will only run on the workers in the joltInit `mainScriptUrlOrBlob` param, ideally a JS file URL. This script can load the Jolt WASM module, and do any special worker setup you need prior to the Jolt physics startup.

This means that we have a solid block where between when threads are created ( initJolt) and when they are run (new JoltInterface).

The kind-of-zero communication logic that you can use (and I have in the example below) is that you can overwrite the WebIDL cache with a JS Proxy, meaning that any arbitrary request for a C++ Pointer will return whatever JS Interface code you want every time.  This technique uses mangled names if the library is distro-mode due to minification, but there are ways around that to prevent them from being minified away, using the -post-js emscripten compiled parameter.

If this proxy concept wasn't so hacky, that could be handled with a custom setClassProxy method to the Wasm Module Interface.

I'm still not solid on a good way to pass data worker thread, mainly pointers, with emscripten's internal scoping of everything Worker related. I was tempted to use BroadcastChannel, but there's not a good way to barrier until all 16 workers are running then send the data and then barrier until they are done. I'd tried working through LocalStorage, BroadcastChannel, IndexedDB ... none are exactly clean, and the solid cleanest way is recoding the Worker JS to add a new `handleMessage` code and pass your arguments and pointers there prior to the actual Thread 'run' command

Note: The code on this branch is * not * the way you'd want to actually do this approach, I just wanted a temp branch to show a way that gets you 80% of the way there. Even if you did plan to just proxy all JS Listeners of a specific class, you'd probably still want to an actual new JS file instead of dynamic JS-blob of text.
https://github.com/PhoenixIllusion/JoltPhysics.js/blob/feature/multi-threaded-contact-listener/Examples/conveyor_belt.worker.html

**Reply by PhoenixIllusion** [reply]
Worst case, we tack on a static `void *` blob worth of memory to a static variable on a custom SharedMem class, for each thread to use after boots up...

**Comment by LeXXik** [comment]
I am not sure if it is of help, but want to share my setup, where I run Jolt in a webworker. I do allow JS callbacks, but without going to main thread. The main concept is that on main thread I stringify the JS function that is supposed to become a callback, send to worker, and evaluate it to turn it back into an executable JS function.

```js
// main thread

function onContactAdded(b1Pointer, b2Pointer, manifoldPointer, settingsPointer) {
    // custom code
}

const overrides = {
    onContactAdded: new String(onContactAdded),
    onContactValidate: ...
};

// send overrides to worker
```

```js
// worker
listener.OnContactAdded = eval('(' + funcStr + ')').bind(this);
```

This way the worker can have a custom JS callback, but without a need to talk to main thread. This is where the User Data fields become useful on the objects. Since these callback have no access to main thread or the app scope, they can use User Data fields to toggle a branch, e.g.:

```js
function onContactSolve(character, ...) {
    ...
    const allowSliding = character.GetShape().GetUserData() === 1 ? true : false;
    if (!allowSliding && ...) {
       ...
    }
}
```

**Reply by PhoenixIllusion** [reply]
What is your worker setup? My current limit is this the passing of anything given the worker is launched deep in the emscripten code and the worker pointers, `onmessage` and `postMessage` are managed in those generated files.

Is this Jolt running in actual multithreading compiled mode, or running single threaded but executing in its own worker, or possibly a combo (shared memory, single thread compile, be careful)

**Reply by LeXXik** [reply]
I see. Sounds like a pain to access them then :( Yeah, my setup is simple - single thread, executed in own worker (non-emscripten worker).

**Comment by PhoenixIllusion** [comment]
iOS 17.4 now appears to be public. This has the Safari fix that allows for multi-threading, so multithreading is theoretically now available on iPhones. I've updated and confirmed the 16-thread-pool build (enable_multi_threading) works on my phone.

The challenge would be how to filter so that iOS 17.3 devices do not attempt to load the multithread builds and fail.

++ Edit: (note, these don't call out anything WASM-Multhreading-Related)  
++ Blog on Release: https://webkit.org/blog/15063/webkit-features-in-safari-17-4/
++ Safari Release Notes: https://developer.apple.com/documentation/safari-release-notes/safari-17_4-release-notes#JavaScript

**Reply by PhoenixIllusion** [reply]
These two are the main requirements for SharedBuffer in general
window.crossOriginIsolated
window.SharedArrayBuffer

User-Agent sniff? It's a horrible thing to do, but I'm not sure how to use the recommended (feature-test) for the Safari bug:
`(/([0-9.]+) Safari/).exec(navigator.userAgent)[1]`
note: Chrome tests as version 122 in this

**Reply by PhoenixIllusion** [reply]
I have initial work on recoding all JS dependent examples to work in multi-threading. Most of them do a no-op or just duplicate blindly the logic of a callback in the worker thread. The one that actually needs to signal back is the buoyancy demo that needs to inform the main thread if the ball is in the moving or still water, so that uses the SharedMemory.

I currently do not have the SharedMemory doing more than storing a fixed value, but the intention was for it to store a pointer to a memory block that all threads could collaborate in.

https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/multi-thread-examples

**Reply by DennisSmolek** [reply]
Do you have a resource for the updates to this safari version?
Perhaps there is something else we can test for besides userAgent

**Reply by LeXXik** [reply]
Not sure there is a reliable way to detect it. Apple is doing all sorts of nonsense to avoid webpages tracking a user via specific hardware fingerprints. Usually what worked were detecting some major features, e.g. if `window.indexedDB` exists, then you are at least on iOS 8. Perhaps there is something that could be used for mapping 17.4 specifically?

https://developer.apple.com/documentation/safari-release-notes/safari-17_4-release-notes

**Reply by DennisSmolek** [reply]
[According to the spec](https://developer.apple.com/documentation/safari-release-notes/safari-17_4-release-notes#JavaScript) they added support for "`detached`" on `arrayBuffers`:
>Added support for ArrayBuffer.prototype.detached, ArrayBuffer.prototype.transfer, and ArrayBuffer.prototype.transferToFixedLength. (118037759)


So this would work:
```js
const arrayBuffer = new ArrayBuffer(8);
const detachedExists = 'detached' in arrayBuffer;
console.log('detached exists', detachedExists);
```

**Reply by PhoenixIllusion** [reply]
Prototype checks are usually enough vs allocating and checking: 
```javascript
!!ArrayBuffer.prototype.detached
```

That specific case will fail on Firefox, which hasn't added `detached` . Feature-Detection is ideally used to check for a feature that is used or required for the desired task. The challenge is the [ideal] solution would be if we had a set of micro-code WAT (text based WASM) of just the offending code that reproduced the issue and could be run for 3-4 iterations in a fraction of a second and reproduce the issue, although even then most of those WAT-based checks confirm success/failure of compilation in general. We know that the code code fails even in single-thread. Ex: Google's WASM feature detectors - https://github.com/GoogleChromeLabs/wasm-feature-detect/tree/main/src/detectors/


There are edge cases (likely not on ArrayBuffer) where checking prototype vs instantiated objects this may cause bugs, like caused by prototype pollution, or special cases where instantiating a class may have special checks in the constructor that forcefully reject/delete/freeze attempts to modify, but those are usually only in the sensitive security sections of the browser or other libraries.

**Comment by PhoenixIllusion** [comment]
If example updates are desired, I'm not sure how we would want to update the examples to support this new build variant

I believe the required update is very minor for most cases, mainly just making sure anything important tags the Bodies with a UserData, but I do not know if it would be confusing to have it mixed in the middle of single-threaded versions. In the best cases, it is just having an optional worker-loader JS file, and you can blindly pass that URL to the initJolt() call and it will never be used unless the code kicks off workers.

The three most complicated ones are:
1) Buoyancy - this involves signaling back to the main thread that the ball has entered a water-field so that it can post-physics-update apply the buoyancy impulse. I currently used a SharedMemory unsigned-long for this.
2) Contact Listener - this one has lots of logging, so the main issue is just how to send all the data back to the main-thread in general to show on the log
3) Multiple Paths - SVG Path Constraint - this one is just a lot of data that would need to be handled by the other threads (multiple JS paths). It is also one of the cases where the SharedMemory may need to be used to distinguish what Path the CPP layer is asking for based on a pointer. The pathing implementation in general for the constraint logic is not good in general. I'd brainstormed it and got distracted and wrote an SVG path-data parser-to-ThreeJS-curve tool, but didn't go much further than that: https://github.com/PhoenixIllusion/JoltPhysics.js/blob/a998f5dd38116fe84547e3d8a0e870a2122e529f/Examples/constraint_svg_path.html#L332 
One technological solution was updating the JS-path logic to use the ThreeCurves, then encoding those curves into shared memory via control-points on the curves, where shared-mem points to control points and workers reconstruct as needed.
Hack-ish solution is that Jolt's PathConstraintPathHermite in CPP can implement any path, but that is kind of just burying the JS-worker issue under the rug.
4) For the non-SVG Path Constraint, I actually got around to implementing it how I'd originally wanted it using just pure Trig logic, so put that trig in the workers in another branch and it worked as expected

**Reply by jrouwe** [reply]
I think it's nice if the examples would support multi threading, but the changes look quite intrusive in some cases and make the simple examples a whole lot less readable/understandable.

I also don't really like this added interface:

```
interface SharedMemory {
	static attribute unsigned long long mPointer;
};
```

It feels like that doesn't belong in the general Jolt interface but should somehow be done on the client side by maybe its own SharedMemory block?.

One possible way forward would be to be able to specify if you want worker threads at initialization time and limit the current examples that use callbacks to just use 1 thread (unless it is trivial to support multiple threads). We could then have one or more examples specifically about handling events on other threads.

**Reply by PhoenixIllusion** [reply]
I do not care for that SharedMemory interface either. Technically a work around is something like local storage as well. Something that is shared between all threads and instantly available, although LocalStorage is also not very clean, and is shared across the entire website domain, not just one page.
I do not understand what is meant by making the change client side.

The issue is back to one of the limits currently with the Emscripten Pool thread spawn logic, and challenges in passing to already spawned Worker threads in the current emscripten system given how locked down it makes the JS Worker message interface. The variable-info would need to be communicated PostJoltInit and either by modification of the emscripten worker.js generated file, or through some other workarounds.

If the JoltPhysics JobSystem can be flagged at runtime to not use the threads, that will at least prevent the examples from crashing. The the builds would still crash if they are a multithreaded compile and not on a COOP server due to window.SharedAreayBuffer being undefined, and if it is defined the pool will still spawn 16 threads for the thread pool if that was part of the compile, even if you only ask for 1 at runtime.

The static Shared value was just a hack-ish compile-time constant built into the library of a safe location to read/write across threads, such as a pointer to a main-thread malloc in the shared array buffer, where there was otherwise no way to share that number post-jolt-init. I had thought about over-allocating the SharedBuffer and pointing past the compile-time ram size for a free swap area, but that is also hacky and gets wiped on a resizable memory compile if you pressure the ram before the workers get a chance to read it.

**Reply by jrouwe** [reply]
> I do not understand what is meant by making the change client side.

This just meant to do it somehow on the JavaScript side in the examples instead of it being part of the library.

**Reply by PhoenixIllusion** [reply]
An ugly method, but one that is mentioned in https://github.com/emscripten-core/emscripten/pull/20235
This method does get a SharedArrayBuffer over to the worker threads and storable in global scope. It breaks if you don't use emscripten Thread Pools and rely on the on-demand pthread creation, but so does WebIDL in general unless you force the usage of PThread.allocateUnusedWorker().

Technically, if you don't have the Thread Pool, you would use [allocate unused workers] to create the pool on-demand yourself with that method. 

To access all the threads in the pool, I added another post-js method:
```javascript
if(Module['PThread']) {
  Module['PThread']['getAllWorkers'] = () => ([... PThread.unusedWorkers, ... PThread.runningWorkers]);
}
```

Then in the worker, used the pattern in the Pull Request that several of their pThread issues said was a 'standard' way to daisy-chain worker startups:
```javascript
// extra *.worker.js file, passed to { mainScriptUrlOrBlob: string }
import initJolt from '../jolt-physics.wasm-compat.js';

export default async (module) => {
  const Jolt = await initJolt(module);

  const handleMessage = globalThis.onmessage;
  const messageQueue = [];
  await new Promise(resolve => {
    function continueLoading() {
      for (let msg of messageQueue) {
        handleMessage(msg);
      }
      globalThis.onmessage = handleMessage;
      resolve();
    }
    globalThis.onmessage = (msg) => {
      if(msg.data.cmd == 'js-shared') {
        const { sharedBuffer, index} = msg.data;
        // TODO: use or store SharedArrayBuffer as needed
        continueLoading();
      } else {
        messageQueue.push(msg);
      }
      if(msg.data.cmd == 'run') {
        continueLoading();
      }
    }
  })

  return Jolt
}
```

Then prior to Jolt Interface constructor, you can pass the SharedArrayBuffer to the children threads. They may not all load before the first call to `JoltInterface.Step()`, but since postMessage is a real queueu, it will queue up and be guaranteed to execute in the order `cmds: ['load', 'js-shared', 'run' ]` (or in that code, optionally skip 'js-shared' if it gets 'run')

Disabling the ThreadPool in the compile, but exposing the PThread.allocateUnusedWorker() on one hand would mean that the same build will run in Single-Threaded mode on non-COOP, but on the other it also requires another work around.
Possibly a utility code that allows you to pass in the thread-counts you want, and it ensures that the pool grows to  X size before calling the JoltInterface's constructor and telling it to start a JobManager of X threads.

**Reply by jrouwe** [reply]
After staring at this code for 10 mins, I think I get what it does (not sure what `getAllWorkers` is for since you don't use it in your code snippet and I don't think you have submitted these changes anywhere yet). My JS knowledge is severely lacking here, so I'm not really sure what the best way forward is.

Github pages doesn't seem to support COOP/COEP so if we enable multi threading on the demos at https://jrouwe.github.io/JoltPhysics.js/ it won't work (although there seem to be hacky [workarounds](https://stackoverflow.com/questions/68609682/is-there-any-way-to-use-sharedarraybuffer-on-github-pages) for this). This means we may need a solution that allows a single threaded fallback (or we just create a separate build for multithreading vs non-multithreading).

**Reply by PhoenixIllusion** [reply]
I had not posted an example of using the new method yet, no.

That would be something like:
```javascript
import initJolt from './js/jolt-physics.wasm-compat.js';

// point to a script with the awaiting Workers, that will sit and wait for a Shared Buffer
initJolt({ mainScriptUrlOrBlob: './worker/load_shared_array.worker.js' }).then(function (Jolt) {
  // allocate a shared memory buffer
  const sharedBuffer = new SharedArrayBuffer(24);
  // use the new method, with newly exposed direct access to Workers, to post a reference
  // of the SharedArrayBuffer to all of them
  Jolt.PThread.getAllWorkers().forEach((worker, i) => 
    worker.postMessage({cmd: 'js-shared', sharedBuffer, index: i }));
  // This command here will create the PhysyicsSystem, which is an active 'execute thread' and
  // will lock the Workers going forward in the while(true) loop. All threads must at least have a
  // pending postMessage of the SharedArrayBuffer before this or forever not have access to
  // it until you destroy and recreate a new PhysicsSystem
  initExample(Jolt);
  
```

**Reply by jrouwe** [reply]
Ok, I understand. Let me know what you think is the best way to proceed (if at all). 

Perhaps add a new version to the package called `jolt-physics-multithreaded.wasm.js` + an example so that people can choose between ease of programming and performance?

**Comment by DennisSmolek** [comment]
I'm having a little bit of a problem following at the end here, but it seems like we're trying to solve many problems at once.

In the future the library could be responsible for deciding which version of the platform to use but for now leave it to the end user. 

I think that means we need a basic build like @jrouwe mentions `jolt-physics-multithreaded.wasm.js` and at least one additional example that shows how to initialize and capitalize on the multithreaded nature. We can add disclaimers that it's experimental and only for chrome. 

Right now, I have to build my own communication api/channels and run Jolt on a separate worker to not have it block the rendering thread. 
From the way it's described here, when built properly, building my own API explicitly and pushing onto a worker first wont be required, and jolt will automatically spin up the workers/threads as needed.

How do we build a simple (understandably limited) version of the Falling Shapes demo that does that? How can we demonstrate the benefits of being multi-threaded? crank the shapes up to 1,000?

Is that possible?

**Reply by PhoenixIllusion** [reply]
For anything that does not actually have callbacks (setup so that during JoltInterface Step, Jolt is a black box), such as Friction, SoftBody, Rigged, Constraints (non-path), Falling Shapes, Add/Remove, 2D Funnel, Stress Test (possibly best example for this), [non-exhaustive list] there is no JS code change or special instruction needed other than building with the ReadMe flags and ensuring the worker-js is in the JS folder.

Regarding browser notes, although this is experimental, it should work equally on newest Safari 17.4, Chrome, Firefox, and Chromium-powered-Edge, so we would not need to state chrome-only.

The only spike in difficulty is if you require even one callback during the physics step, given it is occurring potentially on any spawned external worker process.
